Johnson のアルゴリズム

udon1206 opened the pull request:
close #94
description 及び 概要を書きました．後にもう少し追記する予定ですが，表現に若干自信がないので，ここでいったんレビューしていただけると助かります．

https://jupiro.github.io/algorithm-encyclopedia/johnson-algorithm#noredirect

Commit:
Johnsonのアルゴリズムの見出し

Commit:
Johnsonのアルゴリズムの見出し

Commit:
fix

Commit:
fix

Commit:
概要を追加

Commit:
概要を追加

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,31 @@
+updated_at:
+algorithm:
+  input: >
+    辺重み $c : E \to \mathbb{R}$ 付き有向グラフ $G = (V, E)$ および頂点 $s \in V$
+  output: >
+    各頂点 $t \in V$ に対し $s$-$t$ 最短路長
```
s は指定しないはずです。同様に「$t \in V$ に対し」→「$(s, t) \in V^2$ に対し」でしょうか？

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,31 @@
+  space_complexity: $O(|V| + |E|)$
+  aliases: []
+  level: yellow
+description:  Johnson のアルゴリズムとは、全点対間最短経路問題を解くアルゴリズムのひとつ。負閉路が存在するなら検出できる。 $O(|V| ^ 2 \log |V| + |V||E|)$ で動く。
```
負閉路が、s, t を指定してから検出できるのか、グラフに 1 つでも負閉路があるとそれで終わってしまうのかのどちらでしょうか？後者なら、負閉路検出は Bellman-Ford の部分なので、負閉路の無いグラフに対して動作するアルゴリズムとしてしまってもよい気もしました。

udon1206 replied:
s, tを指定してからではないはずという認識ですので，負閉路の無いグラフに対して動作するアルゴリズムとします．

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,31 @@
+
+1. $\forall (u, v) \in E, \  \hat{c} (u, v) \geq 0$
+
+再重み付けは、 [Bellman-Ford 法](/bellman-ford)を用いて、$ O(\| V \| \| E \|) $ で可能である。再重み付けされたグラフは、辺重みが非負整数であることから、　[Dijkstra 法](/dijkstra) を用いることができるので、$V$ の各頂点において、フィボナッチヒープを用いた [Dijkstra 法](/dijkstra)をすることで、 $O(\|V\| ^ 2 \log \|V\| + \|V\|\|E\|)$ で全点対間最短経路問題を解くことができる。
```
非負整数 → 非負実数

kmyk commented:
nitpick,optional: あまり関係のない commit がまざっているので [`$ git rebase -i origin/gh-pages`](https://docs.github.com/ja/github/using-git/about-git-rebase) をしておくとよいかもしれません。commit log がきれいになるだけなので、難しそうならしなくてもよいとは思っています。なお rebase をすると `git push --force` しないといけなくなりますが、この形の force push はあまり避ける必要はありません。

(内容のレビューは他の人にまかせます)

Commit:
review反映

Commit:
review反映

jupiro commented:
@kmyk 試みてみましたが，```git push --force```が要求されなかったのでなにか僕がミスってますね．．．

@noshi91 レビューありがとうございます．反映しました．

Commit:
reviewer追加

Commit:
reviewer追加

kmyk commented:
@jupiro `git rebase -i origin/gh-pages` は「どう修正しますか？」というのを聞いてくるので、これに適切に返事をしてあげてください。たとえば `pick P(X)の説明追加` となっているところを `drop P(X)の説明追加` と書き換えるとその commit が履歴から削除されます。また、先に `git fetch` をする必要があるかもしれません。
(最初から gh-pages で作業していても大丈夫です。むしろそういうときのための rebase です)

jupiro commented:
@kmyk うまくいったと思います！ありがとうございます！

kmyk commented:
@jupiro rebase 自体はうまくいっていそうです。

ところで、別の問題が起きてしまっている (このプルリクにたくさん commit が表示されてしまっている) ので、以下のように実行してもう一度 rebase してほしいです。私の指示したコマンドが間違っていたのが原因です

``` console
$ git remote add upstream git@github.com:kmyk/algorithm-encyclopedia.git        (初回のみ。2 回目以降は失敗する)
$ git fetch upstream
$ git rebase upstream/gh-pages              (origin は jupiro/algorithm-encyclopedia のことで、そうでなくて upstream の kmyk/algorithm-encyclopedia に rebase する)
```

jupiro commented:
@kmyk これって今回に関わるcommit以外はdropしたらいい感じですか？

kmyk commented:
@jupiro はい。drop すればよいです。
ただし今回の場合であれば、実質的には同じ内容の別の commit がたくさん含まれている形なので、それらは drop しなくても git が自動でいい感じに処理してくれるかもしれません。

jupiro commented:
@kmyk これでうまくいってそうでしょうか？

kmyk commented:
@jupiro 完璧です。ありがとう

Commit:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

Commit:
#105 の反映

Commit:
#105 の反映

kmyk committed:
authors reviewers の表示を changelog 風にしてもっと細かくする

kmyk committed:
CONTRIBUTING.md の authors と reviewers の部分を更新する

以前は「その記事の authors / reviewers」だったのが「その記事その変更の authors / reviewers」になったのがポイントです。

kmyk committed:
scripts/user-ratings.py と scripts/lint.py を修正する

kmyk committed:
中国剰余定理について @MiSawa と @hos-lyric を reviewers を追加する

kmyk committed:
マラソン関係のアルゴリズムについて @MiSawa を reviewers に追加する

kmyk committed:
再帰下降構文解析の reviewers に @MiSawa を追加する

noshi91 commented:
> 後にもう少し追記する予定ですが

追記を待ってから merge するかもう merge してしまうか、どちらが良いでしょうか？

jupiro commented:
@noshi91 追記してから merge で大丈夫でしょうか？

kmyk committed:
authors や reviewers として名前を乗せることの「利益」「不利益」を明確にする

Commit:
再重みつけ追加

Commit:
再重みつけ追加

Commit:
fix

Commit:
fix

jupiro commented:
@noshi91  追記しました．拙い文章ですが，レビューお願いします．

Commit:
fix

Commit:
fix

Commit:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

jupiro commented:
@kmyk markdownの強制改行でwarningがでてるのですが，数式をディスプレイ数式モードで表示するのってどうしたらいいですか？

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,67 @@
+
+## グラフの再重み付け
+
+$G = (V, E)$ に対して、超頂点 $s$ を導入した、 $G' = (V', E')$, $V' = V \cup \lbrace s \rbrace$, $E' = E \cup \lbrace(s, v) : v \in V \rbrace$ を考える。全ての $v \in V$ に対して、 $c(s, v) = 0$ と定義し、辺重みを拡張する。このとき、$G'$ の各頂点の組 $(u, v) \in V \times V$ の辺重み $c$ における最短経路長は $G$ の 辺重み $c$ における最短経路長と一致する。このことから、 $G'$ における全点対間最短経路問題を解くことで、 $G$ の全点対間最短経路問題を解くことができる。再重み付けをした辺重み $\hat{c}$ を以下のようにすることで、概要で述べた再重み付けの性質を持つことを述べる。
```
> 以下のようにする

`以下のように定義する` の方が口語っぽくなくて良さそうです。

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,67 @@
+
+### 性質1. について
+
+$v _ 0 \in V'$ から $v _ {k - 1} \in V'$ の辺重み $c$ 上での最短路 $p = < v _ 0, v _ 1, \dots, v _ {k - 1}>$ とする。このとき、辺重み $\hat{c}$ 上での $p$ の経路長は、
```
> $p = < v _ 0, v _ 1, \dots, v _ {k - 1}>$

`$p = \langle v _ 0, v _ 1, \dots, v _ {k - 1} \rangle$` で求める表示が得られますか？

noshi91 replied:
p を最短路として導入すると最短路以外の路の重みとの比較が出来ないので、「p = <v_0, v_1, ..., v_{k-1}> を v_0 から v_{k-1} へのパスとする。p の \hat{c} 上での重みは hoge ~ したがって、p が c について最短であることと \hat{c} について最短であることは同値である。」と展開すると自然かなと思いました。

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,67 @@
+\hat{c} (p) = \displaystyle \sum _ {n = 1} ^ {k - 1} \hat{c} (v _ {n - 1}, v _ {n}) = c(p) + \delta (s, v _ 0) - \delta (s, v _ {k - 1})
+$ 
+
+となる。 $\delta (s, v _ 0), \delta (s, v _ {k - 1})$ は $p$ に依存しないので、 $p$ が辺重み $c$ における $v _ 0$ から $v _ {k - 1}$ の最短経路であるならば、 辺重み $\hat{c}$ においても最短経路である。
```
$c$ での最短路ならば $\hat{c}$ での最短路である　より、その逆が欲しい気がします。
概要では同値としているので、ここでも「同値である」としてしまってよいと思います。

noshi91 commented on _algorithms/Johnson-algorithm.md:
```diff
@@ -0,0 +1,67 @@
+
+### 性質2. について
+
+$G$ が負閉路を持たないとき、 $G'$ も負閉路を持たない。このことから、
```
> $G'$ も負閉路を持たない。

辺重みが $c$ と $\hat{c}$ の 2 種類あるので、どちらについてなのかの記述が欲しいです。

kmyk committed:
Merge pull request #106 from kmyk/changelog

authors reviewers 関連の表示を改善し運用方法を明文化する

kmyk committed:
KaTeX が display math に対して動いていなかったのをなんとかする

kmyk committed:
display math でなく inline math になっている場合をエラーにする

kmyk committed:
`<` や `>` ではなく `\langle` や `\rangle` を使うようエラーを出す

kmyk committed:
scripts/lint.py 中のいくつかの警告をエラーに昇格する

kmyk committed:
数式中で `<` や `>` を使っていたときのエラーメッセージを修正する

@noshi91 ありがとう

noshi91 committed:
Merge pull request #108 from kmyk/katex

KaTeX で display math が動いていなかった問題などを修正する

kmyk commented:
@jupiro KaTeX のまわりで壊れていました。修正した (https://github.com/kmyk/algorithm-encyclopedia/pull/108) ので `$$` を使ってください。

Commit:
rename

Commit:
rename

Commit:
fix:newline

Commit:
fix:newline

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

jupiro commented:
https://jupiro.github.io/algorithm-encyclopedia/johnson-algorithm#noredirect

@kmyk 数式がうまく表示されていないのですが，どこがやばそうかわかりますか？

kmyk commented:
@jupiro rebase に失敗していそうです。`$ git fetch https://github.com/kmyk/algorithm-encyclopedia && git rebase FETCH_HEAD` ってして push しなおしてみてください。このプルリクで変更されるファイルの一覧 (https://github.com/kmyk/algorithm-encyclopedia/pull/102/files) が Johnson のアルゴリズムのファイルひとつのみになれば成功です。

Commit:
delete

jupiro commented:
```
First, rewinding head to replay your work on top of it...
Applying: Johnsonのアルゴリズムの見出し
Applying: fix
Applying: 概要を追加
Applying: review反映
Applying: reviewer追加
Applying: #105 の反映
Applying: 再重みつけ追加
Applying: fix
Applying: fix
Applying: rename
Using index info to reconstruct a base tree...
.git/rebase-apply/patch:126: trailing whitespace.
$ 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
Applying: fix:newline
Applying: fix
Applying: Johnsonのアルゴリズムの見出し
Using index info to reconstruct a base tree...
Falling back to patching base and 3-way merge...
Applying: fix
Applying: 概要を追加
Applying: review反映
Applying: reviewer追加
Applying: #105 の反映
Applying: 再重みつけ追加
Applying: fix
Applying: fix
Applying: rename
Using index info to reconstruct a base tree...
.git/rebase-apply/patch:126: trailing whitespace.
$ 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
error: Your local changes to the following files would be overwritten by merge:
	_algorithms/johnson-algorithm.md
Please commit your changes or stash them before you merge.
Aborting
error: Failed to merge in the changes.
Patch failed at 0022 rename
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

@kmyk 

試みたところなにかしらでエラーがでてることはわかるんですが，これはどう対処したいいですか？

kmyk commented:
@jupiro これはどのような状態でどのようなコマンドを打った結果ですか？
`_algorithms/johnson-algorithm.md` で conflict が起こっていそう (error: Your local changes to the following files would be overwritten by / merge: _algorithms/johnson-algorithm.md / Please commit your changes or stash them before you merge.) なので、`_algorithms/johnson-algorithm.md` を編集して commit してから `$ git rebase --continue` とするとよいと思います。

---

だめそうなら、以下のようにコマンドを順に実行すると解決すると思います。それでもだめなら (どのようなコマンドを打ったかの部分を含めて) ログの全体を張ってください。

``` console
$ git status
$ git rebase --abort                   # 現在実行中の rebase を中断する
$ git status
$ git stash                            # 現在編集中のファイルを別の場所に退避させる
$ git status                           # 現在の git の状態を表示する
$ git fetch https://github.com/kmyk/algorithm-encyclopedia && git rebase
```

jupiro commented:
@kmyk リポジトリごと削除してやりなおします

このプルリクは一度Closeしたほうがいい感じですかね？

fork しなおして，やりなおしてるのですが，```$$```  の数式表示がうまくいってなさそうです．

kmyk commented:
@jupiro 了解です。
このプルリクについては、リポジトリを削除するときに自動的に close されるんじゃないかなと思います。

ところで、それでも `$$` がうまく表示されないなら、表示されないまま merge してしまってその後に私がなんとかするという選択肢もあります。
